dev/fattest.simplicity/src/componenttest/topology/impl/LibertyServer.java

```diff
@@ -1653,6 +1657,44 @@ protected String validateAppsLoaded(List<String> appList, int intendedTimeout, i
         }
     }
 
+    private void validateTestedFeatures(RemoteFile serverLog) throws Exception {
+        final String m = "validateTestedFeatures";
+        // Load the tested feature data, if it exists
+        File testedFeaturesFile = new File("fat-feature-deps.json");
+        if (!testedFeaturesFile.exists()) {
+            Log.info(c, m, "No tested feature data for this server.  Skipping feature validation");
+            return;
+        }
+
+        // Scrape messages.log to see what features were installed
+        List<String> installedFeaturesRaw = LibertyFileManager.findStringsInFile("CWWKF0012I: .*", serverLog);
+        if (installedFeaturesRaw == null || installedFeaturesRaw.size() == 0)
+            return;
+        Set<String> installedFeatures = new HashSet<String>();
+        for (String f : installedFeaturesRaw)
+            for (String installedFeature : f.substring(0, f.lastIndexOf(']')).substring(f.lastIndexOf('[') + 1).split(","))
+                installedFeatures.add(installedFeature.trim().toLowerCase());
+        Log.info(c, m, "Installed features are: " + installedFeatures);
+
+        // Make sure that any features installed in the server are defined in the fat-feature-deps list
+        JsonArray testedFeaturesJson = Json.createReader(new FileInputStream(testedFeaturesFile)).readArray();
+        Set<String> testedFeatures = new HashSet<String>();
+        testedFeatures.add("timedexit-1.0"); // Manually add timedexit because it's included from an external location
+        for (JsonValue testedFeature : testedFeaturesJson)
+            testedFeatures.add(((JsonString) testedFeature).getString().trim().toLowerCase());
+        for (String installedFeature : installedFeatures) {
+            if (installedFeature.startsWith("usr:"))
+                continue; // Don't need to validate user features
+            if (!testedFeatures.contains(installedFeature))
+                throw new Exception("Installed feature '" + installedFeature +
```

Comment:
Minor enhancement suggestion: Instead of immediately throwing the exception, perhaps build up a list of all of these features that are found and then throw an exception once the loop finishes. That way we get the entire list of features that were installed but not specified as being tested and avoid the need to fix one feature, re-run, find the next failure, and repeat.

Reply:
Good suggestion, I'll make that change